generator/internal/api/visitors.go

```diff
@@ -0,0 +1,316 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import "fmt"
+
+// Element is an interface for all visitable API elements.
+type Element interface {
+	Accept(v *Visitor) error
+}
+
+// Visitor is an interface for visiting API elements, with each visitable element having its own Visit* method.
+//
+// Each node is visited before its children. Parent/child hierarchy is documented in api.API.
+//
+// Traversing the API with a visitor, even the NoOpVisitor, has the side effect of assigning the *.Parent and *.API
+// fields to the visited elements. This side effect is idempotent.
+//
+// Note that as of now, not all API elements are visitable.
+// This was done mostly because there hasn't been a use case for it yet.
+type Visitor interface {
+	VisitAPI(a *API) error
+	VisitMessage(m *Message) error
+	VisitField(f *Field) error
+	VisitOneOf(o *OneOf) error
+	VisitEnum(e *Enum) error
+	VisitService(s *Service) error
+	VisitMethod(m *Method) error
+	VisitPathInfo(p *PathInfo) error
+	VisitPathSegment(s *PathSegment) error
+}
+
+func (a *API) Accept(v Visitor) error {
+	err := v.VisitAPI(a)
+	if err != nil {
+		return err
+	}
+
+	// First we visit the local types, as they are direct children of the API being visited.
+	for _, m := range a.Messages {
+		m.API = a
+		err = m.Accept(v)
+		if err != nil {
+			return err
+		}
+	}
+	// Then we visit the mixin types, as the easiest way to distinguish
+	// them from the local types is that their API field should remain nil.
+	for _, m := range a.State.MessageByID {
```

Comment:
This is visiting the local messages twice.  probably fine if all visitors are idempotent, but maybe a bit wasteful?

Reply:
it loops through, but skips the local elements as they have `Message.API = nil` (guaranteed by the previous loop).